core/util/logging.js

```diff
@@ -10,47 +10,69 @@
  * Logging/debug routines
  */
 
-let _log_level = 'warn';
-
-let Debug = () => {};
-let Info = () => {};
-let Warn = () => {};
-let Error = () => {};
-
-export function init_logging(level) {
-    if (typeof level === 'undefined') {
-        level = _log_level;
-    } else {
-        _log_level = level;
+export const LogLevels = {
+    none: 'none',
+    debug: 'debug',
+    info: 'info',
+    warn: 'warn',
+    error: 'error'
+};
+
+export class  NoopLogger {
+    Debug() {}
+    Info() {}
+    Warn() {}
+    Error() {}
+}
+
+export class ConsoleLogger {
```

Comment:
I'd rather not invent something new here. I think we should use the already standardised [console API](https://developer.mozilla.org/en-US/docs/Web/API/console) as the interface.

Reply:
Do you me the interface or this particular implementation?

I agree with the interface (it's just lowercasing the methods). But the console API is large. We probably shouldn't use the table, time, etc, methods.

If you refer to `ConsoleLogger` in particular. The good thing is that `ConsoleLogger` takes a log level and doesn't log stuff below that level. Unfortunately, you can't do that with the standard console API.

Reply:
Hm, the downside to merging this would be to effectively restrict what kind of logging functions we can use in the internal modules, since we wouldn't want to break an external API once defined.

If we would use the console API we wouldn't at least limit ourselves functionality-wise compared to what's available today.

Reply:
Can you clarify a bit?

The interface that is defined is a subset of the console API i.e. it exposes the commonly used methods of the Console API but not all the methods.

```ts
interface Logger {
  debug(...args) {}
  info(...args) {}
  warn(...args) {}
  error(...args) {}
}
```

If you would like to use more methods from the console API (let's say table) you would just need to add ti and wouldn't break anything.


```ts
interface Logger {
  debug(...args) {}
  info(...args) {}
  warn(...args) {}
  error(...args) {}
  table(...args) {}
}
```

or did you mean something else?

Reply:
We should be nice to people who include our library in their own projects. I mean let's say we initially say that our API requires `debug`, `info`, `warn` & `error`. We then in a later version want to use `table` in our internal modules - that would break the upgrade for everyone who used their own logging module. It would force them to also support `table`.

Reply:
I understand your point.

According to MDN, not including non-standard methods we have:
```
assert()
clear()
count()
countReset()
dir()
dirxml()
error()
group()
groupCollapsed()
groupEnd()
info()
log()
table()
time()
timeEnd()
timeLog()
trace()
warn()
```

So we are missing:
```
assert()
clear()
count()
countReset()
dir()
dirxml()
group()
groupCollapsed()
groupEnd()
log()
table()
time()
timeEnd()
timeLog()
trace()
```

Some of those are not widely supported. For example, `countReset()` or `table()` are not supported by IE11, `timeLog()` is only supported in Firefox, etc.

So the question is: Which of those should be supported?
Does it make sense to support every method?
What happens when Console API add new methods?
Current methods (+ trace) easily work with a logging level. But all the others don't. So how would that work?

Reply:
Generally I guess we wouldn't consider using a function that wasn't supported by all the major browsers.
But yeah I'm starting to realize that I'm probably changing my mind here. Somewhere this feature will grow to a point where the work and maintenance required exceeds the value. I didn't really research what I was asking for.

Also, given @CendioOssman 's intial scepticism I'm leaning towards saying no to this feature. @CendioOssman is MIA for ~1 month at the moment, so it's not much point in waiting for him to give more input.